lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -104,6 +106,78 @@ static bool isZeroBitFIRRTLType(Type type) {
   return type.cast<FIRRTLType>().getPassiveType().getBitWidthOrSentinel() == 0;
 }
 
+namespace {
+struct FirMemory {
+  size_t numReadPorts;
+  size_t numWritePorts;
+  size_t numReadWritePorts;
+  size_t dataWidth;
+  size_t depth;
+  size_t readLatency;
+  size_t writeLatency;
+  bool operator<(const FirMemory &rhs) const {
+#define cmp3way(name)                                                          \
+  if (name < rhs.name)                                                         \
+    return true;                                                               \
+  if (name > rhs.name)                                                         \
+    return false;
+    cmp3way(numReadPorts);
+    cmp3way(numWritePorts);
+    cmp3way(numReadWritePorts);
+    cmp3way(dataWidth);
+    cmp3way(depth);
+    cmp3way(readLatency);
+    cmp3way(writeLatency);
+    return false;
+#undef cmp3way
+  }
+};
+} // namespace
+
+static std::string getFirMemoryName(const FirMemory &mem) {
+  return llvm::formatv("FIRRTLMem_{0}_{1}_{2}_{3}_{4}_{5}_{6}",
+                       mem.numReadPorts, mem.numWritePorts,
+                       mem.numReadWritePorts, mem.dataWidth, mem.depth,
+                       mem.readLatency, mem.writeLatency);
+}
+
+static FirMemory analyzeMemOp(MemOp op) {
+  size_t numReadPorts = 0;
+  size_t numWritePorts = 0;
+  size_t numReadWritePorts = 0;
+
+  for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+    auto portKind = *op.getPortKind(i);
+    if (portKind == MemOp::PortKind::Read)
+      ++numReadPorts;
+    else if (portKind == MemOp::PortKind::Write)
+      ++numReadPorts;
+    else
+      ++numReadWritePorts;
+  }
+  auto width = op.getDataType().getBitWidthOrSentinel();
+  if (width <= 0) {
+    op.emitError("'firrtl.mem' should have simple type and known width");
+    width = 0;
+  }
```

Comment:
Is this check actually looking at `<= 0`, i.e., "memory has unknown width (`-1`) or zero width"? What does "simple type" mean?

Reply:
simple is non struct or vector.  LowerTypes needs to lower memories with complex data types.  #888  This will also give general write mask behavior.

Reply:
But, it continues with zero width as due to the subfield accesses later, it is easier to lower degenerate memories and emit an error than to not lower them and emit an error.

Reply:
In FIRRTL-lingo, "non-struct or vector" would be "ground type" as opposed to "aggregate type".